### PR TITLE
[ConstraintSystem] Switch solver to use `ASTNode` instead of `TypedNode`

### PR DESCRIPTION
**Pull request by xedin**
This is mostly a refactoring to avoid using special typealias `TypedNode` and switch
to use `ASTNode` to represent constraint locator anchor and cache types in the constraint
system. This also allows to drop `const` qualifiers from `{set, get}Type` and align
everything to use `Expr *`.

Note that `{castTo, getAs, is}Expr` accessors still left in place for now, they could be adjusted
separately.

### CODE REVIEW

**Comment by xedin** [comment]
@swift-ci please smoke test

**Review by hamishknight — COMMENTED** [review]
Looks great!

File: lib/Sema/ConstraintLocator.h [hunk 1]
```diff
@@ -384,12 +381,7 @@ class ConstraintLocator : public llvm::FoldingSetNode {
   bool isForOptionalTry() const;
 
   /// Determine whether this locator points directly to a given expression.
-  template <class E> bool directlyAt() const {
-    if (auto *anchor = getAnchor().dyn_cast<const Expr *>()) {
-      return isa<E>(anchor) && getPath().empty();
-    }
-    return false;
-  }
+  template <typename E> bool directlyAt() const;
```
[the remark this answers is not in the document]
Comment (xedin) [reply lib/Sema/ConstraintLocator.h:384]: I tried to replace `is` in `ASTNode` with templated version which uses isa checks instead but that cause a linker failing in swift syntax library so I'd have to investigate separately...

[the remark this answers is not in the document]
Comment (xedin) [reply lib/Sema/ConstraintLocator.h:384]: I'm going to pull implementation into the header for now.

[the remark this answers is not in the document]
Comment (xedin) [reply lib/Sema/ConstraintLocator.h]: Done!

**Comment by xedin** [comment]
@swift-ci please smoke test

**Review by hamishknight — APPROVED** [review]

**Comment by xedin** [comment]
@swift-ci please smoke test

**Comment by xedin** [comment]
@swift-ci please smoke test Windows platform

**Comment by xedin** [comment]
@swift-ci please smoke test macOS platform

**Comment by xedin** [comment]
@swift-ci please smoke test

**Review by hborla — APPROVED** [review]

**Comment by xedin** [comment]
@swift-ci please smoke test

**Comment by xedin** [comment]
@swift-ci please smoke test

**Pull request merged**